File: photomap-core-api/api-gateway/src/main/java/ua/in/photomap/apigateway/util/AuthUtils.java
package ua.in.photomap.apigateway.util;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.DecodedJWT;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.stereotype.Component;

import java.util.Optional;

@Component
@Slf4j
public final class AuthUtils {
    private static final String AUTH_HEADER = "Authorization";
    private static final String AUTH_HEADER_PREFIX = "Bearer ";
    private static final String ACCESS_TOKEN_COOKIE = "accessToken";

    @Value("${photomap.security.secret}")
    private String jwtSecret;

    public Optional<String> getAuthToken(ServerHttpRequest request) {
        return getAuthHeaderToken(request)
                .or(() -> CookieUtils.getCookieValue(request, ACCESS_TOKEN_COOKIE));
    }

    private Optional<String> getAuthHeaderToken(ServerHttpRequest request) {
        return request.getHeaders().getOrEmpty(AUTH_HEADER)
                .stream()
                .filter(header -> header.startsWith(AUTH_HEADER_PREFIX))
                .findFirst()
                .map(header -> header.substring(AUTH_HEADER_PREFIX.length()));
    }

    public DecodedJWT getDecodedJWT(String token) {
        try {
            return JWT.require(Algorithm.HMAC512(jwtSecret))
                    .build()
                    .verify(token);
        } catch (Exception e) {
            log.warn("Failed to decode JWT token: {}", e.getMessage());
            throw e;
        }
    }
}

File: photomap-core-api/api-gateway-kt/src/main/kotlin/ua/in/photomap/apigateway/util/AuthUtils.kt
package ua.`in`.photomap.apigateway.util

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.interfaces.DecodedJWT
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.server.reactive.ServerHttpRequest
import org.springframework.stereotype.Component

@Component
class AuthUtils {
    companion object {
        private val log = LoggerFactory.getLogger(AuthUtils::class.java)
        private const val AUTH_HEADER = "Authorization"
        private const val AUTH_HEADER_PREFIX = "Bearer "
        private const val ACCESS_TOKEN_COOKIE = "accessToken"
    }

    @Value("\${photomap.security.secret}")
    private lateinit var jwtSecret: String

    fun getAuthToken(request: ServerHttpRequest): String? =
        getAuthHeaderToken(request) ?: CookieUtils.getCookieValue(request, ACCESS_TOKEN_COOKIE)

    fun getDecodedJWT(token: String): DecodedJWT = try {
        JWT.require(Algorithm.HMAC512(jwtSecret))
            .build()
            .verify(token)
    } catch (e: Exception) {
        log.warn("Failed to decode JWT token: {}", e.message)
        throw e
    }

    private fun getAuthHeaderToken(request: ServerHttpRequest): String? =
        request.headers.getFirst(AUTH_HEADER)
            ?.takeIf { it.startsWith(AUTH_HEADER_PREFIX) }
            ?.substring(AUTH_HEADER_PREFIX.length)
} 